File: frontend/src/types/payment.ts
export interface PaymentItem {
  name: string
  quantity: number
  price: number
}

export interface CreatePaymentRequest {
  plan_id: string
  billing_cycle: 'monthly' | 'yearly'
  return_url?: string
  cancel_url?: string
}

export interface PaymentLinkResponse {
  order_id: string
  payos_order_code: number
  checkout_url: string
  qr_code: string  // Base64 QR code image
  amount: number
  description: string
}

export interface PaymentStatusResponse {
  order_id: string
  status: 'pending' | 'paid' | 'failed' | 'canceled'
  paid_at: string | null
  payos_transaction_id: string | null
}

export interface Order {
  id: string
  user_id?: string
  order_type: 'subscription' | 'addon'
  subscription_id?: string | null
  amount: number
  status: 'PENDING' | 'PAID' | 'FAILED' | 'CANCELED'
  paid_at: string | null
  is_active?: boolean
  payos_order_code: number | null
  payos_transaction_id?: string | null
  payment_link_id?: string | null
  qr_code?: string | null
  checkout_url?: string | null
  billing_cycle: string | null
  plan_code: string | null
  created_at: string
  updated_at?: string
}

export interface PaymentHistory {
  data: Order[]
  total: number
  limit: number
  offset: number
}

export interface Invoice {
  id: string
  invoice_number: string
  billing_name: string
  billing_address: string
  amount: number
  currency: string
  issue_date: string
  status: 'DRAFT' | 'PAID' | 'CANCELLED'
}

export interface PlanInfo {
  name: string
  code: string
  description: string | null
  monthly_credits: number | null
  available_project: number | null
}

export interface InvoiceDetail {
  invoice: Invoice
  order: Order
  plan: PlanInfo | null
}

File: backend/app/schemas/payment.py
"""Payment schemas for PayOS integration"""

from pydantic import BaseModel, Field
from uuid import UUID
from datetime import datetime
from typing import Optional, Literal


class PaymentItemData(BaseModel):
    """PayOS Item data schema"""
    name: str
    quantity: int = 1
    price: int  # In VND


class CreatePaymentRequest(BaseModel):
    """Request to create payment order"""
    plan_id: UUID
    billing_cycle: Literal["monthly", "yearly"] = "monthly"
    return_url: Optional[str] = None
    cancel_url: Optional[str] = None


class PaymentLinkResponse(BaseModel):
    """Response containing payment link details"""
    order_id: UUID
    payos_order_code: int
    checkout_url: str
    qr_code: str
    amount: int
    description: str


class PaymentStatusResponse(BaseModel):
    """Payment status check response"""
    order_id: UUID
    status: str  # pending, paid, failed, canceled
    paid_at: Optional[datetime] = None
    payos_transaction_id: Optional[str] = None


class PayOSWebhookData(BaseModel):
    """PayOS webhook payload"""
    orderCode: int
    amount: int
    description: str
    accountNumber: str
    reference: str
    transactionDateTime: str
    currency: str = "VND"
    paymentLinkId: str
    code: str
    desc: str
    counterAccountBankId: Optional[str] = None
    counterAccountBankName: Optional[str] = None
    counterAccountName: Optional[str] = None
    counterAccountNumber: Optional[str] = None
    virtualAccountName: Optional[str] = None
    virtualAccountNumber: Optional[str] = None


class WebhookRequest(BaseModel):
    """Webhook request with signature"""
    data: PayOSWebhookData
    signature: str
